### PR TITLE
[GraphQL] Embedded entities support for mutations

### PR DESCRIPTION
**Pull request by alanpoulain**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1750
| License       | MIT
| Doc PR        |

TODO:
- [x] Unit tests

Add a new InputUnionType to allow using either an IRI or data for relations.
It allows to modify embedded entities or to update a related existing resource (#1363).

### CODE REVIEW

File: src/Bridge/Symfony/Bundle/Resources/config/graphql.xml [hunk 1]
```diff
@@ -77,6 +77,10 @@
             <argument type="service" id="api_platform.item_data_provider" on-invalid="ignore" />
             <argument>%api_platform.allow_plain_identifiers%</argument>
 
+            <call method="setItemNormalizer">
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/graphql.xml]: We don't use setter injection in API Platform. Can you inject it directly in the constructor please?

File: src/GraphQl/Serializer/ItemNormalizer.php [hunk 2]
```diff
@@ -62,11 +65,24 @@ public function supportsNormalization($data, $format = null)
         return self::FORMAT === $format && parent::supportsNormalization($data, $format);
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function denormalize($data, $class, $format = null, array $context = [])
+    {
+        return $this->itemNormalizer->denormalize($data, $class, $format, $context);
```
Comment (dunglas) [inline src/GraphQl/Serializer/ItemNormalizer.php]: Can't you use `parent::` directly? It feels weird to me to inject an item normalizer in another one.

File: src/GraphQl/Type/Definition/InputUnionType.php [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+use GraphQL\Utils\Utils;
+
+/**
+ * Represents an union of other input types.
+ *
+ * @experimental
+ *
+ * @author Alan Poulain <contact@alanpoulain.eu>
+ */
+class InputUnionType extends Type implements InputType, LeafType
```
Comment (dunglas) [inline src/GraphQl/Type/Definition/InputUnionType.php]: `final`

File: src/GraphQl/Type/Definition/InputUnionType.php [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+
+    /**
+     * @return InputObjectType[]
+     *
+     * @throws InvariantViolation
+     */
+    public function getTypes(): array
+    {
+        if (null === $this->types) {
```
Comment (dunglas) [inline src/GraphQl/Type/Definition/InputUnionType.php]: I would use a guard clause (`if (null !== $this->types) return...`).

File: src/GraphQl/Type/Definition/InputUnionType.php [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+            if (!isset($this->config['types'])) {
```
Comment (dunglas) [inline src/GraphQl/Type/Definition/InputUnionType.php]:
```php
if ($types = $this->config['types'] ?? null) {
    if (\is_callable($types)) {
        $types = \call_user_func($this->config['types']);
    }
}

File: src/GraphQl/Type/Definition/InputUnionType.php [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    /**
+     * @throws InvariantViolation
+     */
+    public function assertValid()
+    {
+        parent::assertValid();
+
+        $types = $this->getTypes();
+        Utils::invariant(
+            !empty($types),
```
Comment (dunglas) [inline src/GraphQl/Type/Definition/InputUnionType.php]: Can't empty can be removed? You can also inline this.

File: src/GraphQl/Type/Definition/InputUnionType.php [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function parseLiteral($valueNode)
+    {
+        foreach ($this->getTypes() as $type) {
+            if ($type instanceof LeafType) {
+                if (null !== $parsed = $type->parseLiteral($valueNode)) {
```
Comment (dunglas) [inline src/GraphQl/Type/Definition/InputUnionType.php]: can be merged with the previous line:

```php
if ($type instanceof LeafType && null !== $parsed = $type->parseLiteral($valueNode))
```

File: src/Bridge/Symfony/Bundle/Resources/config/graphql.xml (hunk 1 above)
Comment (alanpoulain) [reply src/Bridge/Symfony/Bundle/Resources/config/graphql.xml]: I've done it this way because the constructor of `AbstractItemNormalizer` has a lot of dependencies and I would need to copy it in this normalizer. But I can do it if you prefer.

File: src/GraphQl/Serializer/ItemNormalizer.php (hunk 2 above)
Comment (alanpoulain) [reply src/GraphQl/Serializer/ItemNormalizer.php]: I can't because the default item normalizer is final. And I need it because I don't want to duplicate the code needed to update related existing resources.

File: src/GraphQl/Type/Definition/InputUnionType.php (hunk 6 above)
Comment (alanpoulain) [reply src/GraphQl/Type/Definition/InputUnionType.php]: I've copied it from here: https://github.com/webonyx/graphql-php/blob/master/src/Type/Definition/UnionType.php#L137
I think we can make a `count` instead. Do you agree?

File: src/GraphQl/Serializer/ItemNormalizer.php (hunk 2 above)
Comment (dunglas) [reply src/GraphQl/Serializer/ItemNormalizer.php]: Remove the `final` keyword and add a `@final` PHPDoc annotation then. It will be cleaner IMHO.

**Comment by dunglas** [comment]
It should target master, it's a new feature.

File: src/GraphQl/Type/Definition/InputUnionType.php [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+{
+    /**
+     * @var InputObjectType[]
+     */
+    private $types;
+
+    /**
+     * @var array
+     */
+    public $config;
```
Comment (meyerbaptiste) [inline src/GraphQl/Type/Definition/InputUnionType.php]: `private`?

File: src/Bridge/Symfony/Bundle/Resources/config/graphql.xml (hunk 1 above)
Comment (alanpoulain) [reply src/Bridge/Symfony/Bundle/Resources/config/graphql.xml]: WDYT @dunglas?

Comment (dunglas) [reply src/Bridge/Symfony/Bundle/Resources/config/graphql.xml]: Yes I prefer to copy them. Please :)

**Comment by alanpoulain** [comment]
@dunglas @meyerbaptiste Fixed.

File: src/GraphQl/Serializer/ItemNormalizer.php [hunk 9]
```diff
@@ -39,7 +40,7 @@
      */
     public function normalize($object, $format = null, array $context = [])
     {
-        $data = parent::normalize($object, $format, $context);
+        $data = AbstractItemNormalizer::normalize($object, $format, $context);
```
Comment (meyerbaptiste) [inline src/GraphQl/Serializer/ItemNormalizer.php]: Why this change?

File: src/GraphQl/Type/Definition/InputUnionType.php [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        foreach ($this->getTypes() as $type) {
+            if ($type instanceof LeafType) {
+                try {
+                    return $type->serialize($value);
+                } catch (\Exception $e) {
+                }
+            }
+        }
+
+        throw new InvariantViolation(sprintf('Types in union cannot represent value: %s', Utils::printSafe($value)));
```
Comment (meyerbaptiste) [inline src/GraphQl/Type/Definition/InputUnionType.php]: `@throws` is missing.

File: src/GraphQl/Type/Definition/InputUnionType.php [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        foreach ($this->getTypes() as $type) {
+            if ($type instanceof LeafType) {
+                try {
+                    return $type->parseValue($value);
+                } catch (\Exception $e) {
+                }
+            }
+        }
+
+        throw new Error(sprintf('Types in union cannot represent value: %s', Utils::printSafeJson($value)));
```
Comment (meyerbaptiste) [inline src/GraphQl/Type/Definition/InputUnionType.php]: `@throws` is missing.

File: src/GraphQl/Type/Definition/InputUnionType.php [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+            throw new InvariantViolation(
+                "{$this->name} types must be an Array or a callable which returns an Array."
+            );
+        }
+
+        return $this->types = $types;
+    }
+
+    /**
+     * @throws InvariantViolation
```
Comment (meyerbaptiste) [inline src/GraphQl/Type/Definition/InputUnionType.php]: Should be replaced by `{@inheritdoc}`.

**Comment by alanpoulain** [comment]
@meyerbaptiste Done.

**Comment by dunglas** [comment]
Thank you very much @alanpoulain

File: src/GraphQl/Type/Definition/InputUnionType.php [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+{
+    /**
+     * @var InputObjectType[]
+     */
+    private $types;
+
+    /**
+     * @var array
+     */
+    private $config;
```
Comment (nmoral) [inline src/GraphQl/Type/Definition/InputUnionType.php]: in GraphQL\Type\Definition\Type class $config is declared as public and private in InputUnionType. 

This throw an fatal error with php7.1 and below.

**Comment by Siregacvek** [comment]
@alanpoulain let me ask you why this fix was not merged in production branch?

**Comment by alanpoulain** [comment]
Explanation about why this has been removed is here: https://github.com/api-platform/core/pull/1886

**Comment by Siregacvek** [comment]
@alanpoulain thanks. Looks like it's not solved yet.
